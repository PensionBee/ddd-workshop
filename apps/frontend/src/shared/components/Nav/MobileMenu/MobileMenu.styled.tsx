import styled from "styled-components";

export const StyledContentList = styled("ul")`
  position: relative;
  display: flex;
  flex-direction: column;
  padding: 22px;
  padding-left: unset;
  padding-right: unset;
  margin: 0px;
  column-gap: 10px;
  list-style: none;
  flex: 1;
`;
