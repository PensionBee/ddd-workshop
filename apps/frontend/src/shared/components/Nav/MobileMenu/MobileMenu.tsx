import type { NavMenuItem } from "../Nav.types";
import { StyledContentList } from "./MobileMenu.styled";
import { MobileMenuSubList } from "./MobileMenuSubList/MobileMenuSubList";

export type MobileMenuProps = {
  subList: NavMenuItem["subList"];
};

export const MobileMenu: React.FC<MobileMenuProps> = ({ subList }) => {
  return (
    <StyledContentList>
      <MobileMenuSubList subList={subList} />
    </StyledContentList>
  );
};

export type MobileMainMenuProps = {
  list: NavMenuItem[];
};

export const MobileMainMenu: React.FC<MobileMainMenuProps> = ({ list }) => (
  <MobileMenu subList={list} />
);
