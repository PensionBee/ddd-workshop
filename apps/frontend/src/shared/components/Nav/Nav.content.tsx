import type { NavMenuItem } from "./Nav.types";

// Example content used in tests and stories

export const VISIBLE_LINK: NavMenuItem = {
  children: "Visible Item",
  href: "/",
};
export const HIDDEN_LINK: NavMenuItem = {
  children: "Hidden Item",
  href: "/",
  hidden: true,
};
export const SUB_LIST_ITEM: NavMenuItem = {
  children: "SubList Item",
  subList: [VISIBLE_LINK],
};
export const HIDDEN_SUB_LIST_ITEM: NavMenuItem = {
  children: "Hidden SubList Item",
  subList: [VISIBLE_LINK],
  hidden: true,
};

export const LIST = [VISIBLE_LINK] satisfies NavMenuItem[];
export const LIST_WITH_HIDDEN_LINK = [HIDDEN_LINK] satisfies NavMenuItem[];
export const LIST_WITH_SUB_LIST_ITEM = [SUB_LIST_ITEM] satisfies NavMenuItem[];
export const LIST_WITH_MIXED_ITEMS = [
  VISIBLE_LINK,
  HIDDEN_LINK,
  SUB_LIST_ITEM,
  HIDDEN_SUB_LIST_ITEM,
] satisfies NavMenuItem[];
