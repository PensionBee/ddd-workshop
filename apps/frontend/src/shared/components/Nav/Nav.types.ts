import type { LargeSystemIcon } from "@shared/components/Svg";

export type NavMenuItem = {
  children: string;
  href?: string;
  subList?: NavMenuItem[];
  hidden?: boolean;
  icon?: React.ComponentProps<typeof LargeSystemIcon>["icon"];
};
