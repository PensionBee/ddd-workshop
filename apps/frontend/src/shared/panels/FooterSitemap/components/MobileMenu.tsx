import { GridContainer, GridItem, GridRow } from "@shared/components/Grid";
import { BodySmallL, Category } from "@shared/components/Typography";
import { useState } from "react";
import { Link } from "@shared/components/Link";
import { Box } from "@shared/components/Box";
import { StyledAccordionButton, StyledCaret } from "../FooterSitemap.styled";
import type { MenuLink, MobileMenuList, Menu } from "../FooterSitemap.types";

export const MobileLinkItem: React.FC<MenuLink> = ({
  linkText,
  href,
  target,
  ...props
}) => {
  return (
    <Box as="li" d="flex" pt={10} pb={10} justify="center" {...props}>
      <BodySmallL align="center">
        <Link href={href} target={target} color="black">
          {linkText}
        </Link>
      </BodySmallL>
    </Box>
  );
};

export const MobileAccordion: React.FC<MobileMenuList> = ({
  title,
  list,
  borderTop,
  onClick,
  open,
  ...props
}) => {
  return (
    <GridRow data-state={open ? "open" : "closed"} {...props}>
      <GridItem align="center">
        {borderTop && <Box borderColor="grey4" borderTop={1} w="100%" />}
        <StyledAccordionButton onClick={onClick}>
          <Category>{title}</Category>
          <StyledCaret icon="ChevronYellow" aria-hidden />
        </StyledAccordionButton>
        {open && (
          <Box as="ul" mb={24}>
            {list.map((props) => (
              <MobileLinkItem key={props.linkText} {...props} />
            ))}
          </Box>
        )}
        <Box borderColor="grey4" borderBottom={1} w="100%" />
      </GridItem>
    </GridRow>
  );
};

export const MobileMenu: React.FC<Menu> = ({ sections, ...props }) => {
  const [open, setOpen] = useState<string | null>(null);

  const handleSetOpen = (title: string) => {
    return setOpen((open) => {
      if (title === open) return null;
      return title;
    });
  };

  return (
    <GridContainer pt={{ xs: 40 }} {...props}>
      {Object.entries(sections).map(([key, value], index) => (
        <MobileAccordion
          key={key}
          title={key}
          list={value}
          onClick={() => handleSetOpen(key)}
          open={open === key}
          borderTop={index === 0}
        />
      ))}
    </GridContainer>
  );
};
