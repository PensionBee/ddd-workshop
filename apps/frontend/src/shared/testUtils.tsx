import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
