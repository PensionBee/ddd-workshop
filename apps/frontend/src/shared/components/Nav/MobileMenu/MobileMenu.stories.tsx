import type { ComponentStory, ComponentMeta } from "@storybook/react";
import { LIST_WITH_MIXED_ITEMS } from "../Nav.content";
import { MemoryRouter } from "react-router-dom";
import { Box } from "@shared/components/Box";
import type { MobileMenuProps } from "./MobileMenu";
import { MobileMenu } from "./MobileMenu";

export default {
  title: "Shared/Menus/Mobile Menu",
  component: MobileMenu,
  parameters: {
    viewMode: "story",
    previewTabs: {
      "storybook/docs/panel": { hidden: true },
    },
  },
} as ComponentMeta<typeof MobileMenu>;

const MobileMenuTemplate: ComponentStory<typeof MobileMenu> = (
  args: MobileMenuProps
) => {
  return (
    <MemoryRouter initialEntries={[{ pathname: "/example" }]}>
      <Box d="flex" direction="column" w={320}>
        <MobileMenu {...args} />
      </Box>
    </MemoryRouter>
  );
};

export const Default = MobileMenuTemplate.bind({});
Default.args = {
  subList: LIST_WITH_MIXED_ITEMS,
};
