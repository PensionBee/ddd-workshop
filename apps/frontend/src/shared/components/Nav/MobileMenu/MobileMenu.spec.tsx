import { routerRender } from "@shared/testUtils";
import { MobileMenu } from "./MobileMenu";
import { LIST, LIST_WITH_MIXED_ITEMS } from "../Nav.content";

describe("MobileMenu", () => {
  it("should render correctly", () => {
    const { queryByText, asFragment } = routerRender(
      <MobileMenu subList={LIST} />
    );

    expect(queryByText(LIST[0].children)).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should not render hidden sub list items", () => {
    const { queryByText, asFragment } = routerRender(
      <MobileMenu subList={LIST_WITH_MIXED_ITEMS} />
    );

    LIST_WITH_MIXED_ITEMS.forEach((item) => {
      if (item.hidden) {
        expect(queryByText(item.children)).toBe(null);
      } else {
        expect(queryByText(item.children)).toBeVisible();
      }
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
