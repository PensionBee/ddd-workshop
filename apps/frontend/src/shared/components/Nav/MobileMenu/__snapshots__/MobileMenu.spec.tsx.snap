// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileMenu should not render hidden sub list items 1`] = `
<DocumentFragment>
  <ul
    class="sc-bcXHqe jaBEXw"
  >
    <li
      class="sc-gswNZR"
      data-rc="MobileMenuItem"
    >
      <a
        class="sc-dkrFOg cdXgbd"
        href="/"
      >
        Visible Item
      </a>
    </li>
    <li
      class="sc-gswNZR"
    >
      <div
        class="sc-hLBbgP"
      >
        <button
          class="sc-eDvSVe hLcQzC"
        >
          SubList Item
          <span
            aria-hidden="true"
            class="sc-iBYQkv jlReuQ"
            data-rc="LargeSystemIcon.ArrowRight"
          >
            <svg>
              arrow-right.svg
            </svg>
          </span>
        </button>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`MobileMenu should render correctly 1`] = `
<DocumentFragment>
  <ul
    class="sc-bcXHqe jaBEXw"
  >
    <li
      class="sc-gswNZR"
      data-rc="MobileMenuItem"
    >
      <a
        class="sc-dkrFOg cdXgbd"
        href="/"
      >
        Visible Item
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
